Support for ssl

SonicDMG opened the pull request.

SonicDMG committed:
Adding "alpha" code for ssl support

SonicDMG committed:
Updated DSEConfiguration to more properly handle if SSL is enabled
Added some conditions to check if SSL is enabled and only invoke ssloptions in that case
Made tweaks to compose yaml files to include/remove environ vars

SonicDMG committed:
Refined SSL connectivity code and switched to using netty for SSLContext

SonicDMG committed:
Commit to get the latest working state to Adron

SonicDMG committed:
Added helper script to make it easy to create headless versions of KillrVideo as Docker images

SonicDMG committed:
Added some instructions to the headless helper script

SonicDMG committed:
Made direct config changes to docker-compose external/headless-server yamls for changes that I cannot push to killrvideo-docker-common repo
Once the needed changes are properly versioned and in place in killrvideo-docker-common I will go ahead and git subtree pull them back into the project and back these out

Commit:
Fixins for the cert fixins.

amandamoran commented on lib/killrvideo-docker-common/docker-compose-external.yaml:
```diff
@@ -54,20 +57,28 @@ services:
+      KILLRVIDEO_ENABLE_SSL: $KILLRVIDEO_ENABLE_SSL
       KILLRVIDEO_CASSANDRA_REPLICATION: $KILLRVIDEO_CASSANDRA_REPLICATION
       KILLRVIDEO_LOGGING_LEVEL: $KILLRVIDEO_LOGGING_LEVEL
+      KILLRVIDEO_YOUTUBE_API_KEY: $KILLRVIDEO_YOUTUBE_API_KEY
```
What will KILLRVIDEO_YOUTUBE_API_KEY be used for?

SonicDMG replied:
The youtube key is used to work with their back-end API.  This is how we grab video content from YouTube and display it on KillrVideo.

SonicDMG replied:
You can see this in action if you take a look at the generator service.

jeffreyscarpenter commented on create-headless-server.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Since all this really does is create a docker image, why not name it something like that. Ideally, we should emulate how we build (or even publish) docker images for other repos. See for example the scripts/docker-build.* files under killrvideo-generator: https://github.com/KillrVideo/killrvideo-generator/tree/master/scripts. 

In fact, starting to think we should just go ahead and generate docker images for all of the service implementations, because why not? We should probably do this work in other PRs.

SonicDMG replied:
Hrmmm...Yea, good point on the naming.  To your second point about generating docker images for all the service implementations unless I am misunderstanding your question we already do this.  Each are generated and then stored up on Docker Hub.  These are then pulled down when running docker-compose, no need to generate them locally.  Killrvideo-java is the only one we don't do this with mainly because the "headless" version was really something only we use for our cloud deployment.  I never really intended that to be publicly available as the intent is for folks to run Killrvideo-java via their IDE from a dev perspective.

jeffreyscarpenter commented on lib/killrvideo-docker-common/docker-compose-external.yaml:
```diff
@@ -29,7 +29,7 @@ services:
 
   # Start the KillrVideo web UI on port 3000
   web:
-    image: killrvideo/killrvideo-web:1.2.8
```
good catch, thanks

jeffreyscarpenter commented on src/main/java/killrvideo/configuration/DseConfiguration.java:
```diff
@@ -129,9 +145,55 @@ private void populateContactPoints(Builder clusterConfig)  {
+     * @param clusterConfig
+     *      current configuration
+     */
+    private void populateSSL(Builder clusterConfig) {
```
So, just asking - it looks like if someone deploying this has configured SSL to be on, but misconfigured the certificate file in some way, that this catches the error and keeps going, meaning that the app will keep running without SSL. Is this the best behavior? I guess if the server is configured to require SSL, any calls from this client will fail?

SonicDMG replied:
Right, at this point you won't be able to execute any commands against the database. Yea, I guess we could simply exit out at this point.

clun replied:
Exiting with something like IllegalArgument, this is what i would to 
```java
} catch (FileNotFoundException fne) {
  String errMsg = "SSL cert file not found. You must provide a valid certification file when using SSL encryption option.";
  LOGGER.error(errMsg, fne);
  throw new IllegalArgumentException(errMsg, fne);

} catch (CertificateException ce) {
   String errCert = "Your CA certificate looks invalid. You must provide a valid certification file when using SSL encryption option.";
  LOGGER.error(errCert, ce);
  throw new IllegalArgumentException(errCert, ce);

} catch (Exception e) {
  String errSsl = "Exception in SSL configuration: ";
  LOGGER.error(errSsl, e);
  throw new IllegalArgumentException(errSsl, e);
}
```

SonicDMG replied:
Hrmmm, so @clun, we are not handling anything or responding programmatically if we throw IllegalArgumentException. We want the application to exit on any of these errors.  Why not just exit?

SonicDMG replied:
One other thought.  I've already caught and logged the exception, why throw another one?

SonicDMG replied:
Ok, I pushed some new changes that cleans things up.  I tested the following cases:
no cert file
cert file header broken
cert file footer broken
cert file key incorrect
correct cert file in place

The only one that threw an IllegalArgumentException was when I broke the key.  The others threw the specific exceptions I am explicitly handling for. Stack traces were also properly passed down through the exitOnError() method so they don't get swallowed. I would rather not throw an IllegalArgumentException for things that aren't actually that exception and I think we should definitely exit the application if any of these occur at least so it doesn't look like the app is still running when without the correct cert in place it can't do anything anyway.

clun replied:
Oh you are right, no big change. I am not used to `System.exit` and here are the reasons :

- `System.exit` may be dangerous as it is a violent kill, it doesn't even try to go to `finally` block if exists.  Here it is not a big deal but elsewhere in the code you want to execute some `dseSession.getCluster().close()` at shutdown because the driver is stateful and let sockets opened.

- Here both solutions will quit the application immediately. (As the bean definition is in a `@Configuration` class, if you cannot instanciate the `DseSession` object you will quit immediately). If you throw an exception you got the stacktrace, the line the error occured and a trace in System.err and System.out. As you explicitly LOG an error when the logger and provides the parent exception it the same i guess.

clun replied:
You caught a `checked` Exception but throw a `RuntimeException` (unchecked) to let propagate and stop the application. To my point of view checked exception should be dedicated to business errors - only my feeelings, not a proof whatsoever

SonicDMG replied:
You...are absolutely correct on leaving sockets opened.  I think I can handle that in my exit method before I call System.exit.  Looking into it.

SonicDMG replied:
Ooooooook, @clun, so, there are multiple scenarios.  I went and added a @PreDestroy destroy() method to handle the case of the application being shutdown normally.  In the case I explicitly close the cluster connection per your comments.  I also checked this against the DSE documentation to ensure closing the cluster as compared to the session was the way to go.  It is from what I read.  When closing the cluster it will handle the session as well.  So, there is that.

Now on to the other part and how this all started. For the exceptions we are referencing above I cannot actually close the cluster connection because at the point of failure it has not been established yet.  When I tested this and pointed code to destroy() I would get NullPointerExceptions when trying to access the cluster object to close it, again, because it has not been established yet.  With that, I feel we are safe as far as System.exit() goes.

All in all I effectively have 2 "exit" methods.  I'll admit, that seems strange, but they are for different scenarios. Now that I think of it I guess I could test to see if the cluster connection is established in destroy and choose which path from there.  Thoughts?

clun replied:
Looks good to me.

jeffreyscarpenter commented on src/main/java/killrvideo/service/RatingsService.java:
```diff
@@ -41,7 +41,6 @@
 import killrvideo.validation.KillrVideoInputValidator;
 
 @Service
-//public class RatingsService extends AbstractRatingsService {
```
Does this have to do with a GRPC update? Wondering how it got mixed into this PR.

clun replied:
It is related to GRPC update from 1.0 to 1.10, not sure why it is here in comments

SonicDMG replied:
Neither was I, but since it was not being used I went an removed it.

clun replied:
Makes senses to remove all that, now it's working properly with Grpc 1.10

jeffreyscarpenter reviewed (commented):
Let's add a page to the killrvideo.github.io site to document this SSL configuration and our security options in general.

SonicDMG committed:
Made changes per pull request #33
Renamed create-headless-server.sh to create-docker-image.sh to better reflect what the script is actually doing
Updated the SSL portion of DSEConfiguration to exit out of the application for cases where the cert file is either invalid or does not exist.

SonicDMG committed:
Merge branch 'support_for_ssl' of https://github.com/KillrVideo/killrvideo-java into support_for_ssl

Adron reviewed (approved):
LGTM. :100:

clun commented on src/main/java/killrvideo/configuration/DseConfiguration.java:
```diff
@@ -57,6 +66,12 @@
+    @Value("#{environment.KILLRVIDEO_ENABLE_SSL}")
+    public Optional < Boolean > dseEnableSSL;
+
+    @Value("${killrvideo.ssl.CACertFileLocation: 'cassandra.cert'}")
```
I propose to use the key `killrvideo.cassandra.ssl.CACertFileLocation` in the `application.yaml` to identify that the SSL connectivity is associated with DSE.

SonicDMG replied:
Good point, done

clun commented on src/main/java/killrvideo/configuration/DseConfiguration.java:
```diff
@@ -129,9 +145,59 @@ private void populateContactPoints(Builder clusterConfig)  {
+     *      current configuration
+     */
+    private void populateSSL(Builder clusterConfig) {
+        if (dseEnableSSL.isPresent() &&  dseEnableSSL.get()) {
```
Idea: `if (dseEnableSSL.isPresent() &&  dseEnableSSL.get()) {` could be directly moved in caller do not invoke populateSSL at all if the flag is not enabled.

SonicDMG replied:
Yea, makes sense

SonicDMG replied:
Are you thinking the same thing with the populateAuthentication() method then as well?  Otherwise we'll essentially have 2 different implementation styles in place.

clun replied:
Yes makes sense to do it with `populateAuthentication()` as well, sure. 👍will do when merged

SonicDMG replied:
Ok, taking a look now.

SonicDMG replied:
Ok, did some refactoring here.

clun commented on src/main/java/killrvideo/configuration/DseConfiguration.java:
```diff
@@ -146,7 +212,7 @@ private void populateAuthentication(Builder clusterConfig) {
             clusterConfig.withAuthProvider(cassandraAuthProvider);
             String obfuscatedPassword = new String(new char[dsePassword.get().length()]).replace("\0", "*");
-            LOGGER.info(" + Using supplied DSE username: '%s' and password: '%s' from environment variables", 
+            LOGGER.info(" + Using supplied DSE username: '{}' and password: '{}' from environment variables",
```
Thanks, ok I did not test properly obviously when did that in the first place

clun reviewed (approved):
Few tips here and there. I integrated those changes in v2 and works as expected

SonicDMG committed:
Made another set of changes per pull request https://github.com/KillrVideo/killrvideo-java/pull/33
Updated both the README.md and pom.xml to use the latest DSE Java 1.6.8 driver
Updated local docker-compose.yaml and lib/killrvideo-docker-common/docker-compose.yaml to reference 2.1.0 versions of dse-config, generator, and webUI
Updated key killrvideo.ssl.CACertFileLocation in application.yaml to killrvideo.cassandra.ssl.CACertFileLocation. This way it is clear it is related to cassandra.
Cleaned up some exception handling to now fully exit the application on certain SSL errors.  If SSL is enabled it makes no sense to continue running the app if there is an error in the cert file or it doesn't exist.  This will also ensure the Docker container housing this service exits as well in case of an error otherwise it looks like it is still running.  Cleaned up double logging as well since I was both logging the output and then printing the stack trace.  Just need one.

SonicDMG committed:
Refactored code based off of Cedrick's comments per pull request https://github.com/KillrVideo/killrvideo-java/pull/33
Moved checks against some Optional variables like "if (dseEnableSSL.isPresent() && dseEnableSSL.get()) {" outside of their implementing methods and into their respective callers.  I also switched from using isPresent() and get() to orElse().  Hopefully I did not switch from one anti-pattern to another one, but this is easier to read IMO and prevents a chicken and egg scenario when trying to read a get() before isPresent() is evaluated.

SonicDMG committed:
Did more refactoring in DseConfiguration based off of Cedrick's comments in https://github.com/KillrVideo/killrvideo-java/pull/33 regarding some ideas on properly shutting the application down on exit.

SonicDMG committed:
Updating dse version from 6.0.0 to 6.0.2

SonicDMG committed:
Updated Studio Docker version from 6.0.0 to 6.0.2
